lib/fake_async.dart

```diff
@@ -211,39 +211,50 @@ class FakeAsync {
       {Duration timeout = const Duration(hours: 1),
       bool flushPeriodicTimers = true}) {
     final absoluteTimeout = _elapsed + timeout;
-    _fireTimersWhile((timer) {
-      if (timer._nextCall > absoluteTimeout) {
-        // TODO(nweiz): Make this a [TimeoutException].
-        throw StateError('Exceeded timeout $timeout while flushing timers');
+    for (;;) {
+      // With [flushPeriodicTimers] false, continue firing timers only until
+      // all remaining timers are periodic *and* every periodic timer has had
+      // a chance to run against the final value of [_elapsed].
+      if (!flushPeriodicTimers) {
+        if (_timers
+            .every((timer) => timer.isPeriodic && timer._nextCall > _elapsed)) {
+          break;
+        }
       }
 
-      if (flushPeriodicTimers) return _timers.isNotEmpty;
+      if (!runNextTimer(timeout: absoluteTimeout - _elapsed)) {
+        if (_timers.isEmpty) break;
 
-      // Continue firing timers until the only ones left are periodic *and*
-      // every periodic timer has had a change to run against the final
-      // value of [_elapsed].
-      return _timers
-          .any((timer) => !timer.isPeriodic || timer._nextCall <= _elapsed);
-    });
+        // TODO(nweiz): Make this a [TimeoutException].
+        throw StateError('Exceeded timeout $timeout while flushing timers');
+      }
+    }
   }
 
-  /// Invoke the callback for each timer until [predicate] returns `false` for
-  /// the next timer that would be fired.
+  /// Elapses time to run one timer, if any timer exists.
   ///
-  /// Microtasks are flushed before and after each timer is fired. Before each
-  /// timer fires, [_elapsed] is updated to the appropriate duration.
-  void _fireTimersWhile(bool Function(FakeTimer timer) predicate) {
-    flushMicrotasks();
-    for (;;) {
-      if (_timers.isEmpty) break;
+  /// Microtasks are flushed before and after the timer runs. Before the
+  /// timer runs, [elapsed] is updated to the appropriate value.
+  ///
+  /// The [timeout] controls how much fake time may elapse. If non-null,
+  /// then timers further in the future than the given duration will be ignored.
+  ///
+  /// Returns true if a timer was run, false otherwise.
```

Comment:
Do you have a use case in mind for the return value? This is useful internally, but I can't immediately think of a testing pattern where it would be a good idea.

What do you think about implementing `bool _runNextTimer` for internal usage, and `void runNextTimer` wrapping it to hide the bool until someone describes a good use case?

Reply:
Yep, exactly. Thanks for the review!